### PR TITLE
Adding requirement crumbs for Materials and Composites

### PR DESCRIPTION
**Pull request by john-science**
## What is the change?

Adding requirement crumbs for Materials and Composites.

## Why is the change being made?

This is part of on-going requirement work.

---

## Checklist

- [X] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [X] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify that the new/changed code works.

- [X] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [X] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [X] The [release notes](https://terrapower.github.io/armi/release/index.html) (location `doc/release/0.X.rst`) are up-to-date with any important changes.
- [X] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [X] No [requirements](https://terrapower.github.io/armi/developer/tooling.html#watch-for-requirements) were altered.
- [X] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/materials/tests/test_materials.py [hunk 1]
```diff
@@ -80,6 +80,10 @@ def test_duplicate(self):
         .. test:: Test the material base class.
             :id: T_ARMI_MAT_PROPERTIES3
             :tests: R_ARMI_MAT_PROPERTIES
+
+        .. test:: Materials shall calc mass fracs at init.
+            :id: T_ARMI_MAT_FRACS
+            :tests: R_ARMI_MAT_FRACS
```
Comment (albeanth) [inline armi/materials/tests/test_materials.py:86]: this test is on a protected test class and doesn't show up in the API.

Comment (albeanth) [reply armi/materials/tests/test_materials.py:86]: same with the other test tags here. Either they all need to be moved or the class needs to be made public.

Comment (john-science) [reply armi/materials/tests/test_materials.py:86]: Dammit

Comment (john-science) [reply armi/materials/tests/test_materials.py:86]: Drat

File: armi/reactor/tests/test_composites.py [hunk 2]
```diff
@@ -124,6 +134,16 @@ def test_iterComponents(self):
         self.assertIn(self.thirdGen, list(self.container.iterComponents()))
 
     def test_getChildren(self):
+        """Test the get children method.
+
+        .. test:: Components are a physical part of the reactor.
```
Comment (albeanth) [inline armi/reactor/tests/test_composites.py:139]: Same comment here about the test not actually verifying that components are a part of the reactor.

File: armi/reactor/tests/test_blocks.py [hunk 3]
```diff
@@ -307,6 +306,8 @@ def getComponentData(component):
 
 class TestDetailedNDensUpdate(unittest.TestCase):
     def setUp(self):
+        from armi.reactor.blueprints.tests.test_blockBlueprints import FULL_BP
```
Comment (albeanth) [inline armi/reactor/tests/test_blocks.py:309]: Why move this down here?

Comment (john-science) [reply armi/reactor/tests/test_blocks.py:309]: While testing this PR I found an infinite import loop in ARMI.

It's between `test_blocks.py` and `test_blockBlueprints.py`.

It turns out Python usually figures it's way around it, but I hit a situation (running just certain tests in parallel), where it didn't.

So I fixed the infinite import loop.  Unrelated to crumbs, obviously.

File: armi/reactor/tests/test_composites.py (hunk 2 above)
Comment (john-science) [reply armi/reactor/tests/test_composites.py:139]: I have a test for that part of the req in test_reactors.py:HexReactorTests:test_coreSfp()

File: armi/reactor/components/component.py [hunk 4]
```diff
@@ -256,6 +256,10 @@ def __lt__(self, other):
         True if a circle encompassing this object has a smaller diameter than one encompassing another component.
 
         This allows sorting because the Python sort functions only use this method.
+
+        .. impl:: Order components by there outermost diameter.
+            :id: I_ARMI_COMP_ORDER
+            :implements: R_ARMI_COMP_ORDER
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/components/component.py:262]: Drat. Moved to the class module.  No other choice, really.

File: armi/materials/tests/test_materials.py (hunk 1 above)
Comment (john-science) [reply armi/materials/tests/test_materials.py]: Okay, this private test class had a BUNCH of req crumbs. I moved them all to a public test class on a specific material.

File: armi/reactor/tests/test_composites.py [hunk 5]
```diff
@@ -110,7 +110,17 @@ def setUp(self):
         container.add(nested)
         self.container = container
 
-    def test_Composite(self):
+    def test_composite(self):
+        """Test basic Composite things.
+
+        .. test:: Components are a physical part of the reactor.
```
[the remark this answers is not in the document]
Comment (john-science) [reply armi/reactor/tests/test_composites.py]: Fixed

File: armi/materials/tests/test_uZr.py [hunk 6]
```diff
@@ -13,16 +13,117 @@
...
 
+    def setUp(self):
+        self.mat = self.MAT_CLASS()
+
+    def test_isPicklable(self):
+        """Test that materials are picklable so we can do MPI communication of state.
+
+        .. test:: Test the material base class.
+            :id: T_ARMI_MAT_PROPERTIES0
+            :tests: R_ARMI_MAT_PROPERTIES
```
Comment (albeanth) [inline armi/materials/tests/test_uZr.py:34]: This doesn't feel relevant to the linked requirement. Let's chat offline.

File: armi/materials/tests/test_uZr.py [hunk 7]
```diff
@@ -13,16 +13,117 @@
...
+        self.mat.adjustTD(0.5)
+        self.assertEqual(0.5, self.mat.theoreticalDensityFrac)
+        self.assertEqual(self.mat.cached, {})
+
+    def test_duplicate(self):
+        """Test the material duplication.
+
+        .. test:: Test the material base class.
+            :id: T_ARMI_MAT_PROPERTIES3
+            :tests: R_ARMI_MAT_PROPERTIES
```
Comment (albeanth) [inline armi/materials/tests/test_uZr.py:65]: This doesn't feel relevant to the linked requirement. Let's chat offline.

File: armi/materials/tests/test_uZr.py [hunk 8]
```diff
@@ -13,16 +13,117 @@
...
+        self.assertEqual(mat.parent, self.mat.parent)
+        self.assertEqual(mat.refDens, self.mat.refDens)
+        self.assertEqual(mat.theoreticalDensityFrac, self.mat.theoreticalDensityFrac)
+
+    def test_cache(self):
+        """Test the material cache.
+
+        .. test:: Test the material base class.
+            :id: T_ARMI_MAT_PROPERTIES4
+            :tests: R_ARMI_MAT_PROPERTIES
```
Comment (albeanth) [inline armi/materials/tests/test_uZr.py:86]: This doesn't feel relevant to the linked requirement. Let's chat offline.

File: armi/reactor/tests/test_blocks.py (hunk 3 above)
Comment (albeanth) [reply armi/reactor/tests/test_blocks.py:309]: Huh. Cool! Good stuff. 

Does it make sense to move it to a `setUpClass` class method so the import isn't done for every single test?

Comment (john-science) [reply armi/reactor/tests/test_blocks.py:309]: I mean, that entire class only has ONE test.  So, it's not a big deal.

And that test class isn't imported or subclasses anywhere.

Comment (john-science) [reply armi/reactor/tests/test_blocks.py:309]: In fact, I am just going to remove the `setUp()` method from that class. It's just one test that needs to be independent from others anyway.

File: armi/materials/tests/test_uZr.py (hunk 6 above)
Comment (john-science) [reply armi/materials/tests/test_uZr.py:34]: I updated the comment to show that it does.

File: armi/materials/tests/test_uZr.py (hunk 7 above)
Comment (john-science) [reply armi/materials/tests/test_uZr.py]: I agree, removed.

File: armi/materials/tests/test_uZr.py (hunk 8 above)
Comment (john-science) [reply armi/materials/tests/test_uZr.py]: I agree, removed.

**Review by albeanth — APPROVED** [review]

**Pull request merged**